docs(graphql): Add language tag support (#199)
dgraph-io/dgraph-docs: content/graphql/dgraph/index.md

*Note: the current behavior requires that when two fields are mapped to the same Dgraph predicate both should have the same `@search` directive.  This is likely to change in a future release where the underlying Dgraph indexes will be the union of the `@search` directives, while the generated GraphQL API will expose only the search given for the particular field.  Allowing, for example, dgraph predicate name to have `term` and `hash` indexes, but exposing only term search for GraphQL movies and hash search for GraphQL people.*

## Language tag support in GraphQL

In your GraphQL schema, you need to define different fields for each language tag that you want to use. 
You also need to set the Dgraph predicate name of the untagged field in the `@dgraph` argument of language tag field. 

> Dgraph will automatically add a `@lang` directive in the Dgraph schema for the corresponding predicate.

Reviewer: ```suggestion
Dgraph will automatically add a `@lang` directive in the DQL schema for the corresponding predicate.
```

For example:

```graphql
type Person {
     name: String   // Person.name is the corresponding dgraph field for this field

>      nameHi: String @dgraph(pred:"Person.name@hi")

Reviewer: ```suggestion
     nameHi: String @dgraph(pred:"Person.name@hi") # this field exposes the value for the language tag `@hi` for the DQL predicate `Person.name` to GraphQL
```

     nameEn: String @dgraph(pred:"Person.name@en")

>      nameHi_En:  String @dgraph(pred:"Person.name@hi:en") // won't be added to mutation patch

Reviewer: ```suggestion
     nameHi_En:  String @dgraph(pred:"Person.name@hi:en") # this field uses multiple language tags: `@hi` and `@en`
```

>      nameHi_En_untag:  String @dgraph(pred:"Person.name@hi:en:.") //won't be added to mutation patch

Reviewer: ```suggestion
     nameHi_En_untag:  String @dgraph(pred:"Person.name@hi:en:.") # as this uses `.`, it will give untagged values if there is no value for `@hi` or `@en`
```

  }
```

{{% notice "tip" %}}

> The Dgraph predicate name for the corresponding GraphQL field is defined as `typename.fieldname`.

Reviewer: ```suggestion
By default, the DQL predicate for a GraphQL field is generated as `Typename.FieldName`.
```

{{% /notice %}}

> For example, in `nameHi: String @dgraph(pred:"Person.name@hi")`, the Dgraph predicate for the corresponding field name in GraphQL is `Person.name`.
> If you set the argument with a different field name, then it won’t work as expected.
> 
> {{% notice "note" %}}
> You need to define all language fields at one place, either in Type or in the Interface.
> Dgraph won’t add fields in mutation patch which have multiple language tags.
> For example, field `nameHi_En: String @dgraph(pred:"Person.name@hi:en")` can only be queried.
> {{% /notice %}}

Reviewer: some of it isn't needed, some of it has been covered already in the example and the rest I have rephrased below:
```suggestion
If a GraphQL field uses more than one language tag, then it won't be part of any mutation input. Like, in the above example the fields `nameHi_En` and `nameHi_En_untag` can't be given as an input to any mutation. Only the fields which use one or no language can be given in a mutation input, like `name`, `nameHi`, and `nameEn`.

All the fields can be queried, irrespective of whether they use one language or more.

To know more about language support in DQL, please refer: https://dgraph.io/docs/master/tutorial-4
```

### Limitations

> - Dgraph won’t be able to query `Person.name@*` type of code because it returns a list. Although this can be achieved by querying all the codes.
> - Fields which  corresponding to multiple language tags, such as: 
>     ```graphql
>     nameHiEn: String @dgraph(pred: "Person.name@hi:en"`)
>     nameHi_En_Untag: String @dgraph(pred: "Person.name@hi:en:.")
>     ```
>   are not updatable/mutable, and won’t be added to update/add, mutation/reference type or in any filter, orders type.
>   

Reviewer: ```suggestion
GraphQL won’t be able to query `Person.name@*` type of language tags because of the structural requirements of GraphQL.
```
Also, maybe we can just put this one line as a note/warning instead of adding a sub-heading `Limitations`.

### Caveats

Dgraph automatically adds the `@lang` directive to the untagged language field in the Dgraph schema.
If an untagged language field doesn’t exist for a language tagged field, then Dgraph automatically adds that field in the corresponding Dgraph schema.

For example, consider the following GraphQL schema:

```graphql
type Person {   
     name: String! @id    
     nameHi: String @dgraph(pred: "Person.name@hi") @search(by: [term, exact])  
     professionEn: String @dgraph(pred: "Person.profession@en")
 }
```

The generated Dgraph schema will be:

```graphql
 type Person {     
     Person.name

>      Person.profession 
>  }
> 
> Person.name: string @index(exact, hash, term) @lang @upsert .
> Person.profession: string @lang .
> ```
> 
> It is also possible that an untagged field is defined in the interface and a tagged field is defined in an implementation type. 
> There can also be a case where the type specified in the tagged language field doesn’t exist in the GraphQL schema.
> In that case, Dgraph simply assumes that the DQL schema already has that type and generates the corresponding predicate with a `@lang` predicate.
> 
> For example:
> 
> ```graphql
> interface Node {
>   f1: String
> }
> 
> type Person implements Node {
>     f1Hi: String @dgraph(pred: "Node.f1@hi")
>     f2: String @dgraph(pred: "T.f@no")
> 
>     name: String! @id
> 
> }
> ```
> 
> The generated Dgraph schema:
> 
> ```graphql
> type Node {
>     Node.f1
> }
> 
> Node.f1: string @lang .
> 
> type Person {
>     Node.f1
>     T.f
>     Person.name
> }
> 
> T.f: string @lang .
> Person.name: string @index(hash) @upsert .
> ```
> 
> ### Interaction with exiting directives for language tagged fields
> 
> **Schema directives**
> 
> - `@id`: currently not required.
> - `@search`: applicable only on one field (tagged or untagged value) and will apply on all GraphQL fields which map to the same Dgraph predicate.
> - `@lambda`: doesn’t work with `@dgraph` directive
> - `@custom`: doesn’t work with `@dgraph` directive
> - `@hasinverse`: doesn’t apply on `String` fields
> 
> **Query directives**
> 
> - `@skip`: Work normally as with other fields
> - `@include`: Work normally as with other fields

Reviewer: This shouldn't be required. These are implementation details and not true use-cases, so don't think should be part of the user-facing documentation.
```suggestion
```
